src/layer.cpp

```diff
@@ -214,17 +223,114 @@ scene_on_touch_begin(VALUE rcv, SEL sel)
     }
     block = rb_retain(block); // FIXME need release...
 
-    auto listener = cocos2d::EventListenerTouchOneByOne::create();
-    listener->onTouchBegan = [block](cocos2d::Touch *touch,
+    auto scene = SCENE(rcv);
+    if (!RTEST(scene->touch_listener)) {
```

Comment:
We should do this instead:

```
if (scene->touch_listener == NULL) {
```

The `RTEST` macro is reserved for Ruby objects, and here it's a C++ object pointer.